include/taskolib/Sequence.h

```diff
@@ -290,10 +290,19 @@ class Sequence
      *
      * If the sequence finished normally, the returned string is empty.
      *
-     * \see execute()
+     * \see get_error_step_index(), set_error_message_and_step_index()
      */
     const std::string& get_error_message() const noexcept { return error_message_; }
 
+    /**
+     * Return the index of the step at which the sequence stopped prematurely.
+     *
+     * If the sequence finished normally, this function returns nullopt.
+     *
+     * \see get_error_message(), set_error_message_and_step_index()
+     */
+    OptionalStepIndex get_error_step_index() const { return error_step_index_; }
```

Comment:
Nice 👍 ! But would it be better to have only one member function (for example `get_error()`) that returns a `struct` instead of using two getter functions?

```c++
struct Error
{
    std::string message_;
    OptionalStepIndex index_;
};

...
Error& get_error() { ... }
```

On success we can return `Error{"", std::nullopt}`. I am aware that this is not compatible with older releases...

Reply:
> But I also see that we return the `Error` as `ref` (as before we return the string as `ref`). When I see that I know why C++ has a reputation to be difficult and error prone, as that imposes the responsibility of lifetime management on the user of the lib. And for what? The sake of premature optimization?

This in fact is regarding `string const& Sequence::get_error_message()`:

![image](https://user-images.githubusercontent.com/16012374/206509367-2b293cef-276e-4b6c-8ab9-2afc007a4e54.png)

What about [Core Guideline F.15](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f15-prefer-simple-and-conventional-ways-of-passing-information):

![image](https://user-images.githubusercontent.com/16012374/206510008-b9d1cd62-5c4f-450c-9888-43ad2ec3ae14.png)

Never return a `const&`.

Reply:
Additional rant against `string_view`:

This is actually broken-ish on MSVC, while on Linux the generated code is just forwarding the pointer and length in two registers on the MSVC-ABI the whole `string_view` object will be copied (!) and a pointer to the new `string view` is passed. That is why people on MS use `foo(sting_view const& str)` :-(

Reply:
> This in fact is regarding `string const& Sequence::get_error_message()`:
> [...] 
> What about [Core Guideline F.15](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f15-prefer-simple-and-conventional-ways-of-passing-information):
> [...]
> Never return a `const&`.

I think you have a point for this specific case here (we are talking about an error message of ~100 characters or so), but to my eyes the general advice "never return a const&" is wrong, and is also not stated as such in the core guidelines. In fact, [Core Guideline F.44](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-return-ref) explicitly acknowledges that references can be returned if copies are not desirable.

Consider our beloved DOOCS text property, `D_text`. Internally it stores a `std::string` with the text. The main accessor, `const std::string& get_value() const`, returns a reference to that string. Why? Because the text might be large – we have seen hundreds of kilobytes in `D_text` – and it simply would be unacceptable to copy all that if users just want to test for `prop.get_value().empty()` or search for a substring.

There is the possibility of misusing functions that return references to member variables, sure, but the default usage is perfectly safe:
```
Sequence seq(...);
std::string msg = seq.get_error_message();
```

Of course, as always in C++, if you assign things to reference types or take pointers, you need to be careful and think twice. You are correct about `string_view` in that regard. That does not mean that it cannot be used safely, though. In fact, your example is broken because of wrong usage of `string_view`, not because the return value of `get_message()` is a reference. It would be just as broken if it returned a `std::string`:
![image](https://user-images.githubusercontent.com/28754274/206659345-f1b3e1ed-fc5c-41ae-b160-3aafd77b893c.png)

Reply:
> Additional rant against `string_view`:
> 
> This is actually broken-ish on MSVC, while on Linux the generated code is just forwarding the pointer and length in two registers on the MSVC-ABI the whole `string_view` object will be copied (!) and a pointer to the new `string view` is passed. That is why people on MS use `foo(sting_view const& str)` :-(

I sympathize, but if you worry about copying the whole `string_view` object (a whopping 16 bytes), I do not see how this fits with advocating the copying of the entire error message on each call to `get_error_message()`.

Reply:
> but to my eyes the general advice "never return a const&" is wrong

Well, that is even in my eyes wrong ;-D How would that ever work with operators that return a reference to `this` ;-)

> That does not mean that it cannot be used safely

For me the question is: Can it be easily used unsafe? And string-view is a type that can easily lead people to dangling stuff, at least when exposed, sometimes, etc. Everything can be used safely if all obscure aspects are observed.

What I have see there is

a) The danger that returning a ref spreads (also to functions that should not return that), see MissPP's example, which returned a ref, modelled after `get_error_message()` that returns a ref.

b) Optimization without need. Here it is an error message. That should be never a hot path, right? So why event THINK about any optimizations? Keep the code as simple and unpretentious as possible. Copy the string if it help, that will hardly be any difference. Especially where we throw optimizations in other places away; the string stuff gets too much attention and leads to code less straight forward. Optimize `cat()` :+1: Optimize `set_error_message()` :-1:

;-)

> I do not see how this fits with advocating

That is unrelated (and thus in another comment), just ... if you are in a hot path and want to optimize, than there are unexpected implementation details :-(